### PR TITLE
Fix #2597 by checking that the directory exists

### LINKED ISSUE #2597: Poetry 1.1.0a2 includes python code in the version string
To reproduce:

1. `poetry init` with default choices; no dependencies.
1. `poetry add pytest-cov`
1. `poetry install`

The last command prints:
```
Installing dependencies from lock file

Package operations: 0 installs, 1 update, 0 removals

- Updating pytest-cov (2.10.0 import os, sys;exec('if \'COV_CORE_SOURCE\' in os.environ:\n try:\n  from pytest_cov.embed import init\n  init()\n except Exception as exc:\n  sys.stderr.write(\n   "pytest-cov: Failed to setup subprocess coverage. "\n   "Environ: {0!r} "\n   "Exception: {1!r}\\n".format(\n    dict((k, v) for k, v in os.environ.items() if k.startswith(\'COV_CORE\')),\n    exc\n   )\n  )\n') -> 2.10.0) 
```

### PR DESCRIPTION
**Pull request by PetterS**
Resolves: #2597

- [x] Added **tests** for changed code.

### CODE REVIEW

**Comment by PetterS** [comment]
@abn

**Comment by PetterS** [comment]
Will fix the Python 2.7 errors if this is reasonable.

File: poetry/repositories/installed_repository.py [hunk 1]
```diff
@@ -49,20 +49,19 @@ def load(cls, env):  # type: (Env) -> InstalledRepository
...
-                    ):
-                        with env.site_packages.joinpath(
-                            "{}.pth".format(package.pretty_name)
-                        ).open() as f:
+                    pth_file = env.site_packages.joinpath(
+                        "{}.pth".format(package.pretty_name)
+                    )
+                    if path.name.endswith(".dist-info") and pth_file.exists():
+                        with pth_file.open() as f:
                             directory = Path(f.readline().strip())
```
Comment (abn) [inline poetry/repositories/installed_repository.py:57]:
```suggestion
                    directories = set()
                    site.addpackage(env.site_packages.as_posix(), "{}.pth".format("package.pretty_name"), directories)
                    if directories:
                        # TODO: handle multiple source directories
                        package.source_type = "directory"
                        package.source_url = directories.pop()
                        continue
```

File: poetry/repositories/installed_repository.py [hunk 2]
```diff
@@ -49,20 +49,19 @@ def load(cls, env):  # type: (Env) -> InstalledRepository
...
-                            package.source_url = directory.as_posix()
-
-                    continue
+                            # The PTH file can contain executable code, but we don't execute it
+                            # here and only consider static directories.
+                            # https://github.com/python-poetry/poetry/issues/2597
+                            if directory.exists():
+                                package.source_type = "directory"
+                                package.source_url = directory.as_posix()
+                                continue
```
Comment (abn) [inline poetry/repositories/installed_repository.py:64]:
```suggestion
```

File: poetry/repositories/installed_repository.py (hunk 1 above)
Comment (PetterS) [reply poetry/repositories/installed_repository.py:57]: Where does `site.addpackage` come from? I don't see the string `addpackage` anywhere in the code base.

Comment (abn) [reply poetry/repositories/installed_repository.py:57]:
```
>>> import site
>>> site.addpackage
<function addpackage at 0x7f9a39dfd310>
```

**Comment by PetterS** [comment]
Did the suggested changes. I don't understand the code well enough to write really good tests here, I'm afraid. The test currently mocks `site.addpackage`.

File: tests/repositories/test_installed_repository.py [hunk 3]
```diff
@@ -43,6 +47,7 @@ def test_load(mocker):
             {"remote.origin.url": "git@github.com:sdispater/pendulum.git"},
         ],
     )
+    mocker.patch("site.addpackage", side_effect=addpackage)
```
Comment (abn) [inline tests/repositories/test_installed_repository.py:50]: Unsure why this is required. If code works as expected, the discovered `site-packages` directory would be `tests/repositories/fixtures/installed/lib/python3.7/site-packages/`. This contains a file `editable.pth`, which would be read and hence loading its content `/path/to/editable` as the value for `editable.source_url`.

**Comment by PetterS** [comment]
OK, I reverted the test file so you can see the error.

File: poetry/repositories/installed_repository.py [hunk 4]
```diff
@@ -49,20 +51,17 @@ def load(cls, env):  # type: (Env) -> InstalledRepository
...
-                        ).open() as f:
-                            directory = Path(f.readline().strip())
-                            package.source_type = "directory"
-                            package.source_url = directory.as_posix()
-
-                    continue
+                    directories = set()
+                    site.addpackage(
+                        env.site_packages.as_posix(),
+                        "{}.pth".format("package.pretty_name"),
```
Comment (abn) [inline poetry/repositories/installed_repository.py:57]:
```suggestion
                        "{}.pth".format(package.pretty_name),
```

File: poetry/repositories/installed_repository.py [hunk 5]
```diff
@@ -49,20 +51,17 @@ def load(cls, env):  # type: (Env) -> InstalledRepository
...
+                    site.addpackage(
+                        env.site_packages.as_posix(),
+                        "{}.pth".format(package.pretty_name),
+                        directories,
+                    )
+                    if directories:
+                        # TODO: handle multiple source directories
+                        package.source_type = "directory"
+                        package.source_url = directories.pop()
+                        continue
```
Comment (abn) [inline poetry/repositories/installed_repository.py:64]:
```suggestion
                    try:
                        pth_file = env.site_packages.joinpath("{}.pth".format(package.pretty_name)).resolve(strict=True)
                        with pth_file.open() as f:
                            for line in f:
                                line = line.strip()
                                if not line or line.startswith(("#", "import ", "import\t")):
                                    # we ignore executables and comments
                                    continue
    
                                # we expect this line to be either a relative or an absolute path
                                path = Path(line)
    
                                if not path.is_absolute():
                                    path = env.site_packages.joinpath(path)
    
                                if path.is_dir():
                                    # TODO: handle multiple source directories?
                                    package.source_type = "directory"
                                    package.source_url = path.as_posix()
                                    break
                    except FileNotFoundError:
                        pass
```
It looks like the `addpackage` method has as a side-effect that is undesirable here.

File: poetry/repositories/installed_repository.py [hunk 6]
```diff
@@ -1,3 +1,5 @@
+import site
+
```
Comment (abn) [inline poetry/repositories/installed_repository.py:2]:
```suggestion
```

File: poetry/repositories/installed_repository.py [hunk 7]
```diff
@@ -48,21 +48,29 @@ def load(cls, env):  # type: (Env) -> InstalledRepository
...
+                                if not path.is_absolute():
+                                    path = env.site_packages.joinpath(path)
+    
+                                if path.is_dir():
+                                    # TODO: handle multiple source directories?
+                                    package.source_type = "directory"
+                                    package.source_url = path.as_posix()
+                                    break
+                    except FileNotFoundError:
+                        pass
```
Comment (abn) [inline poetry/repositories/installed_repository.py:73]:
```suggestion
                if is_standard_package and path.name.endswith(".dist-info"):
                    try:
                        pth_file = env.site_packages.joinpath(
                            "{}.pth".format(package.pretty_name)
                        ).resolve(strict=True)
                        with pth_file.open() as f:
                            for line in f:
                                line = line.strip()
                                if not line or line.startswith(
                                    ("#", "import ", "import\t")
                                ):
                                    # we ignore executables and comments
                                    continue

                                # we expect this line to be either a relative or an absolute path
                                path = Path(line)

                                if not path.is_absolute():
                                    path = env.site_packages.joinpath(path)

                                if path.is_dir():
                                    # TODO: handle multiple source directories?
                                    package.source_type = "directory"
                                    package.source_url = path.as_posix()
                                    break
                    except FileNotFoundError:
                        pass
```

File: poetry/repositories/installed_repository.py [hunk 8]
```diff
@@ -10,6 +12,38 @@
```
Comment (abn) [inline poetry/repositories/installed_repository.py:12]:
```suggestion

try:
    FileNotFoundError
except NameError:
    FileNotFoundError = IOError

```

**Comment by abn** [comment]
@PetterS I have applied a few changes. The changes themselves work; but need to rework for compatibility and add test case for executable `.pth` files.

File: poetry/repositories/installed_repository.py [hunk 9]
```diff
@@ -8,8 +10,45 @@
 
 _VENDORS = Path(__file__).parent.parent.joinpath("_vendor")
 
+try:
+    FileNotFoundError
+except NameError:
+    FileNotFoundError = IOError
+
```
Comment (abn) [inline poetry/repositories/installed_repository.py:17]:
```suggestion
```

File: poetry/repositories/installed_repository.py [hunk 10]
```diff
@@ -8,8 +10,45 @@
...
+            with pth_file.open() as f:
+                for line in f:
+                    line = line.strip()
+                    if line and not line.startswith(("#", "import ", "import\t")):
+                        path = Path(line)
+                        if not path.is_absolute():
+                            path = sitedir.joinpath(path)
+                        paths.add(path)
+        except FileNotFoundError:
+            pass
```
Comment (abn) [inline poetry/repositories/installed_repository.py:48]:
```suggestion
        pth_file = sitedir.joinpath("{}.pth".format(name))
        if pth_file.exists():
            pth_file = sitedir.joinpath("{}.pth".format(name)).resolve(strict=True)
            with pth_file.open() as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith(("#", "import ", "import\t")):
                        path = Path(line)
                        if not path.is_absolute():
                            path = sitedir.joinpath(path)
                        paths.add(path)
```

File: poetry/repositories/installed_repository.py [hunk 11]
```diff
@@ -10,6 +12,37 @@
...
+        Reference: https://docs.python.org/3.8/library/site.html
+
+        :param sitedir: The site-packages directory to search for .pth file.
+        :param name: The name of the package to search .pth file for.
+        :return: A `Set` of valid `Path` objects.
+        """
+        paths = set()
+        pth_file = sitedir.joinpath("{}.pth".format(name))
+        if pth_file.exists():
+            pth_file = sitedir.joinpath("{}.pth".format(name)).resolve(strict=True)
```
Comment (abn) [inline poetry/repositories/installed_repository.py:34]:
```suggestion
```

**Comment by PetterS** [comment]
Thanks for fixing this!

**Pull request merged**